File: app/src/main/java/com/dmi/perfectreader/book/BookPresenter.java
package com.dmi.perfectreader.book;

import android.content.Context;

import com.dmi.perfectreader.book.pagebook.PageBook;
import com.dmi.perfectreader.book.pagebook.PageBookRenderer;
import com.dmi.perfectreader.bookstorage.EPUBBookStorage;
import com.dmi.perfectreader.setting.AppSettings;
import com.dmi.perfectreader.userdata.UserData;
import com.dmi.util.base.BasePresenter;
import com.dmi.util.setting.AbstractSettingsApplier;

import java.io.File;
import java.io.IOException;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import timber.log.Timber;

import static com.dmi.util.concurrent.Threads.postIOTask;
import static com.dmi.util.concurrent.Threads.postUITask;
import static com.dmi.util.lang.IntegerPercent.ZERO;
import static com.google.common.base.MoreObjects.firstNonNull;

@Singleton
public class BookPresenter extends BasePresenter {
    @Inject
    @Named("bookFile")
    protected File bookFile;
    @Inject
    @Named("applicationContext")
    protected Context context;
    @Inject
    protected UserData userData;
    @Inject
    protected AppSettings appSettings;
    @Inject
    protected EPUBBookStorage bookStorage;
    @Inject
    protected BookFragment view;

    private final SettingsApplier settingsApplier = new SettingsApplier();
    private PageBook pageBook;
    private TapHandler tapHandler = null;

    @Override
    public void onCreate() {
        settingsApplier.startListen();
        pageBook.goPercent(loadLocation());
        postIOTask(this::loadBook);
    }

    @Override
    public void onDestroy() {
        settingsApplier.stopListen();
    }

    public PageBookRenderer createRenderer() {
        return null;
    }

    protected void loadBook() {
        try {
            bookStorage.load(bookFile);
            postUITask(this::afterBookStorageLoad);
        } catch (IOException e) {
            Timber.e(e, "Load book error");
            view.showBookLoadingError();
        }
    }

    protected void afterBookStorageLoad() {
        settingsApplier.applyAll();
    }

    private int loadLocation() {
        return firstNonNull(userData.loadBookLocation(bookFile), ZERO);
    }

    protected void saveLocation() {
        int currentLocation = 0;
        postIOTask(() -> userData.saveBookLocation(bookFile, currentLocation));
    }

    public void resize(int width, int height) {
        pageBook.resize(width, height);
    }

    public void resume() {
    }

    public void pause() {
    }

    public int currentPercent() {
        return 0;
    }

    public void tap(float x, float y, float tapDiameter, TapHandler tapHandler) {
        this.tapHandler = tapHandler;
        pageBook.tap(x, y, tapDiameter);
    }

    private void handleTap() {
        if (tapHandler != null) {
            tapHandler.handleTap();
            tapHandler = null;
        }
    }

    public void goPercent(int percent) {
        view.reset(() -> {
            pageBook.goPercent(percent);
            saveLocation();
        });
    }

    public void goNextPage() {
        if (pageBook.canGoPage(-view.currentPageRelativeIndex() + 1) != PageBook.CanGoResult.CANNOT) {
            view.goNextPage();
        }
    }

    public void goPreviewPage() {
        if (pageBook.canGoPage(-view.currentPageRelativeIndex() - 1) != PageBook.CanGoResult.CANNOT) {
            view.goPreviewPage();
        }
    }

    public int synchronizeCurrentPage(int currentPageRelativeIndex) {
        if (currentPageRelativeIndex < 0) {
            switch (pageBook.canGoPage(1)) {
                case CAN:
                    pageBook.goNextPage();
                    saveLocation();
                    return currentPageRelativeIndex + 1;
                case CANNOT:
                    return 0;
            }
        } else if (currentPageRelativeIndex > 0) {
            switch (pageBook.canGoPage(-1)) {
                case CAN:
                    pageBook.goPreviewPage();
                    saveLocation();
                    return currentPageRelativeIndex - 1;
                case CANNOT:
                    return 0;
            }
        }
        return currentPageRelativeIndex;
    }

    private class SettingsApplier extends AbstractSettingsApplier {
        public void applyAll() {
        }

        @Override
        protected void listen() {
        }
    }

    public interface TapHandler {
        void handleTap();
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/book/BookPresenter.kt
package com.dmi.perfectreader.book

import android.content.Context
import com.dmi.perfectreader.app.AppThreads.postIOTask
import com.dmi.perfectreader.book.pagebook.PageBook
import com.dmi.perfectreader.book.pagebook.PageBookRenderer
import com.dmi.perfectreader.bookstorage.EPUBBookStorage
import com.dmi.perfectreader.setting.AppSettings
import com.dmi.perfectreader.userdata.UserData
import com.dmi.util.base.BasePresenter
import com.dmi.util.setting.AbstractSettingsApplier
import com.google.common.base.MoreObjects.firstNonNull
import timber.log.Timber
import java.io.File
import java.io.IOException
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

@Singleton
open class BookPresenter : BasePresenter() {
    @Inject
    @Named("bookFile")
    protected lateinit  var bookFile: File
    @Inject
    @Named("applicationContext")
    protected lateinit var context: Context
    @Inject
    protected lateinit var userData: UserData
    @Inject
    protected lateinit var appSettings: AppSettings
    @Inject
    protected lateinit var bookStorage: EPUBBookStorage
    @Inject
    protected lateinit var view: BookFragment

    private val settingsApplier = SettingsApplier()
    private val pageBook: PageBook? = null
    private var tapHandler: TapHandler? = null

    public override fun onCreate() {
        settingsApplier.startListen()
        pageBook!!.goPercent(loadLocation())
        postIOTask { this.loadBook() }
    }

    public override fun onDestroy() {
        settingsApplier.stopListen()
    }

    fun createRenderer(): PageBookRenderer {
        return object: PageBookRenderer {
            override val isLoading: Boolean
                get() = throw UnsupportedOperationException()

            override fun onSurfaceCreated() {
                throw UnsupportedOperationException()
            }

            override fun onSurfaceChanged(width: Int, height: Int) {
                throw UnsupportedOperationException()
            }

            override fun onFreeResources() {
                throw UnsupportedOperationException()
            }

            override fun onDrawFrame() {
                throw UnsupportedOperationException()
            }
        }
    }

    protected fun loadBook() {
        try {
            bookStorage.load(bookFile)
            postIOTask { this.afterBookStorageLoad() }
        } catch (e: IOException) {
            Timber.e(e, "Load book error")
            view.showBookLoadingError()
        }

    }

    protected fun afterBookStorageLoad() {
        settingsApplier.applyAll()
    }

    private fun loadLocation(): Double {
        return firstNonNull(userData.loadBookLocation(bookFile), 0.0)
    }

    protected fun saveLocation() {
        val currentLocation = 0
        postIOTask { userData.saveBookLocation(bookFile, currentLocation.toDouble()) }
    }

    fun resize(width: Int, height: Int) {
        pageBook!!.resize(width, height)
    }

    fun resume() {
    }

    fun pause() {
    }

    open fun currentPercent(): Double {
        return 0.0
    }

    open fun tap(x: Float, y: Float, tapDiameter: Float, tapHandler: TapHandler) {
        this.tapHandler = tapHandler
        pageBook!!.tap(x, y, tapDiameter)
    }

    private fun handleTap() {
        if (tapHandler != null) {
            tapHandler!!.handleTap()
            tapHandler = null
        }
    }

    open fun goPercent(percent: Double) {
        view.reset(Runnable {
            pageBook!!.goPercent(percent)
            saveLocation()
        })
    }

    open fun goNextPage() {
        if (pageBook!!.canGoPage(-view.currentPageRelativeIndex() + 1) !== PageBook.CanGoResult.CANNOT) {
            view.goNextPage()
        }
    }

    open fun goPreviewPage() {
        if (pageBook!!.canGoPage(-view.currentPageRelativeIndex() - 1) !== PageBook.CanGoResult.CANNOT) {
            view.goPreviewPage()
        }
    }

    fun synchronizeCurrentPage(currentPageRelativeIndex: Int): Int {
        if (currentPageRelativeIndex < 0) {
            when (pageBook!!.canGoPage(1)) {
                PageBook.CanGoResult.CAN -> {
                    pageBook.goNextPage()
                    saveLocation()
                    return currentPageRelativeIndex + 1
                }
                PageBook.CanGoResult.CANNOT -> return 0
            }
        } else if (currentPageRelativeIndex > 0) {
            when (pageBook!!.canGoPage(-1)) {
                PageBook.CanGoResult.CAN -> {
                    pageBook.goPreviewPage()
                    saveLocation()
                    return currentPageRelativeIndex - 1
                }
                PageBook.CanGoResult.CANNOT -> return 0
            }
        }
        return currentPageRelativeIndex
    }

    private inner class SettingsApplier : AbstractSettingsApplier() {
        fun applyAll() {
        }

        override fun listen() {
        }
    }

    interface TapHandler {
        fun handleTap()
    }
}
